packages/flutter/lib/src/widgets/modal_button_sheet_route.dart

```diff
@@ -0,0 +1,321 @@
+import 'dart:async';
+import 'dart:ui';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+import 'modal_button_sheet.dart';
+
+// Route that defines the transition animation of the previous route when this
+// one is closing
+mixin RouteWithPreviousTransitionMixin<T> on Route<T> {
+  Widget getPreviousRouteTransition(
+    BuildContext context,
+    Animation<double> secondAnimation,
+    Widget child,
+  );
+}
+
+// Route that allows the next route to define the animation transition of this route
+// when the route appears back after the next one is popped
+mixin AllowsRouteWithPreviousTransitionMixin<T> on PageRoute<T> {
```

Comment:
Right now the back/forth transitions are defined by the route itself. For the modals, the previous route transition needs to be defined by the modal route itself. I am not sure if my approach here is the best way to do it, but this is the only way I managed to do it with the current situation of the framework.

It would be nice to add the modals as the new Page class for the Navigator 2.0